atc/engine/engine.go

```diff
@@ -175,12 +180,20 @@ func (b *engineBuild) Run(logger lager.Logger) {
 
 	defer notifier.Close()
 
-	step, err := b.builder.BuildStep(b.build)
+	step, err := b.builder.BuildStep(b.build, logger)
 	if err != nil {
 		logger.Error("failed-to-build-step", err)
+
+		// Fails the build if BuildStep returned error. Because some unrecoverable error,
+		// like pipeline credential manager is wrong, will cause a build to never start
+		// to run.
+		builder.NewBuildStepDelegate(
```

Comment:
This kinda sticks out like a sore thumb. :thinking: How likely are these errors to occur?

Reply:
Without this feature, `BuildStep` seems to rarely fail though there are a few clauses returning error. But with this feature, if, like Vault config in pipeline yaml is wrong (meaning invalid vault URL, not something can be blocked by `Validate`), then `BuildStep` will fail, this kind of error could be very possible. I think, in any case, as long as `BuildStep` returns error, then error should be handled.

Reply:
I just meant that the constructing of the delegate in-place and calling a method on it is kind of weird looking and probably hard to cover with a test. I agree though, it would be nice to surface to the user. If you could add a unit test for this, which I suspect would involve some refactoring, I'm good with it. :+1:

Maybe the error logging could be moved into the builder somehow since it already has a delegate factory injected? :man_shrugging: Might be some other way to clean it up too.